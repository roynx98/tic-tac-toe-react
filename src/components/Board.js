import React from 'react';

import { Square } from './Square';

export class Board extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      squares: Array(9).fill(null),
      isXNext: true
    };
  }

  handleClick(i) {
    // Make a copy of the array
    const squares = this.state.squares.slice();

    squares[i] = this.state.isXNext ? 'X' : 'O';
    this.setState({
      squares,
      isXNext: !this.state.isXNext
    });
  }

  renderSquare(i) {
    return (
      <Square 
        value={this.state.squares[i]} 
        onClick={() => {this.handleClick(i)}}
        />
    );
  }

  render() {
    const status = `Next player: ${this.state.isXNext ? 'X' : 'O'}`;

    return (
      <div>
        <div className="status">{status}</div>
        <div className="board-row">
          {this.renderSquare(0)}
          {this.renderSquare(1)}
          {this.renderSquare(2)}
        </div>
        <div className="board-row">
          {this.renderSquare(3)}
          {this.renderSquare(4)}
          {this.renderSquare(5)}
        </div>
        <div className="board-row">
          {this.renderSquare(6)}
          {this.renderSquare(7)}
          {this.renderSquare(8)}
        </div>
      </div>
    );
  }
}