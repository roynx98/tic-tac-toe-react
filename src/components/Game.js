import { Board } from './Board';
import React from 'react';
import { calculateWinner } from '../helpers/calculateWinner';

export class Game extends React.Component {

  constructor(props) {
    super(props);

    this.state = {
      history: [{
        squares: Array(9).fill(null)
      }],
      isXNext: false
    };
  }

  handleClick(i) {
    const history = this.state.history;
    const current = history[history.length - 1];

    if (current.squares[i] ||Â calculateWinner(current.squares)) {
      return;
    }

    const newHistory = this.state.history.slice();
    const newSquares = current.squares.slice();

    newSquares[i] = this.state.isXNext ? 'X' : 'O';

    this.setState({
      history: newHistory.concat({ squares: newSquares }),
      isXNext: !this.state.isXNext
    });
  }

  render() {
    const history = this.state.history;
    const current = history[history.length - 1];
    const winner = calculateWinner(current.squares);

    let status = `Next player: ${this.state.isXNext ? 'X' : 'O'}`;
    if (winner) {
      status = `Winner: ${winner}`;
    }

    return (
      <>
        <div className="game">
          <div className="game-board">
            <Board 
              squares={current.squares}
              onClick={i => this.handleClick(i)} />
          </div>
          <div className="game-info">
            <div>{status}</div>
            <ol>{/* TODO */}</ol>
          </div>
        </div>
      </>
    );
  }

}
