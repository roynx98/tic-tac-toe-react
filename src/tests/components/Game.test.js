import React from 'react';
import { shallow } from 'enzyme';
import { Game } from '../../components/Game';

describe('Game component', () => {

  test('should match snapshot', () => {
    const wrapper = shallow(<Game />);
    expect(wrapper).toMatchSnapshot();
  });

});