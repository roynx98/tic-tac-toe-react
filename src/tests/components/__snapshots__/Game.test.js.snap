// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game component should match snapshot 1`] = `
<Fragment>
  <div
    className="game"
  >
    <div
      className="game-board"
    >
      <Board
        onClick={[Function]}
        squares={
          Array [
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
          ]
        }
      />
    </div>
    <div
      className="game-info"
    >
      <div>
        Next player: O
      </div>
      <ol />
    </div>
  </div>
</Fragment>
`;
